packages/effect/src/internal/data.ts

```diff
@@ -34,3 +35,108 @@ export const Structural: new<A>(
 /** @internal */
 export const struct = <As extends Readonly<Record<string, any>>>(as: As): As =>
   Object.assign(Object.create(StructuralPrototype), as)
+
+const deepSymbol = Symbol.for("effect/Data/deep")
+
+/** @internal */
+export const proxy = <A>(value: A, options?: { deep: boolean }): A => {
+  if ((typeof value === "function" || typeof value === "object") && value !== null) {
+    let hashCache: any = "INIT"
+    return new Proxy(value, {
+      has(target, p) {
+        if (p === Hash.symbol || p === Equal.symbol || p === deepSymbol) {
+          return true
+        }
+        return p in target
+      },
+      get(target, p) {
+        if (p === deepSymbol) {
+          return options?.deep
+        }
+        if (p === Hash.symbol) {
+          return () => {
+            if (hashCache === "INIT") {
+              hashCache = options?.deep === true
+                ? deepHash(value)
+                : Array.isArray(value)
+                ? Hash.array(value)
+                : Hash.structure(value)
+            }
+            return hashCache
+          }
+        }
+        if (p === Equal.symbol) {
+          if (options?.deep === true) {
+            return deepComp
+          }
+          if (Array.isArray(value)) {
+            return ArrayProto[Equal.symbol]
+          } else {
+            return StructuralPrototype[Equal.symbol]
+          }
+        }
+        // @ts-expect-error
+        return target[p]
+      }
+    })
+  }
+  return value
+}
+
+function deepHash(value: any) {
+  if (typeof value === "object" && value !== null) {
+    if (Array.isArray(value)) {
+      let h = 6151
+      for (let i = 0; i < value.length; i++) {
+        h = pipe(h, Hash.combine(deepHash(value[i])))
+      }
+      return Hash.optimize(h)
+    } else {
+      const keys = Object.keys(value)
+      let h = 12289
+      for (let i = 0; i < keys.length; i++) {
+        h ^= pipe(Hash.string(keys[i]! as string), Hash.combine(deepHash((value as any)[keys[i]!])))
+      }
+      return Hash.optimize(h)
+    }
+  }
+  return Hash.hash(value)
+}
+
+function deepComp(this: any, that: any) {
+  if (typeof that === typeof this && that !== null && this[deepSymbol] === that[deepSymbol]) {
```

Comment:
```suggestion
  if (typeof that === typeof this && that !== null && this[deepSymbol] && that[deepSymbol]) {
```

Only if both have deep equality enabled?

Reply:
yes, equality should be symmetric no?

Reply:
Just wasn't sure if setting deep to false on both should pass the condition.